homeassistant/components/voip/rtp_audio.py

```diff
@@ -0,0 +1,260 @@
+"""Utility for converting audio to/from RTP + OPUS packets."""
+import audioop
+from collections.abc import Iterable
+from dataclasses import dataclass
+import logging
+import random
+import struct
+from typing import Any
+
+import opuslib
+
+from .error import RtpError
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class RtpOpusInput:
+    """Extracts audio from RTP packets with OPUS."""
+
+    opus_rate: int = 48000  # Hz
+    opus_width: int = 2  # bytes
+    opus_channels: int = 2
+    opus_frame_size: int = 960  # samples per channel
+    opus_payload: int = 123  # set by GrandStream
+
+    def __post_init__(
+        self,
+    ) -> None:
+        """Initialize encoder and state."""
+        self._decoder = opuslib.api.decoder.create_state(
+            self.opus_rate, self.opus_channels
+        )
+
+    def process_packet(
+        self,
+        rtp_bytes: bytes,
+        rate: int,
+        width: int,
+        channels: int,
+    ) -> bytes:
+        """Extract, decode, and return raw audio from RTP packet."""
+        if channels not in (1, 2):
+            raise ValueError("Only mono and stereo audio is supported")
+
+        # Minimum header size
+        if len(rtp_bytes) < 12:
+            raise RtpError("RTP packet is too small")
+
+        # See: https://en.wikipedia.org/wiki/Real-time_Transport_Protocol#Packet_header
+        flags, payload_type, _sequence_num, _timestamp, _ssrc = struct.unpack(
+            ">BBHLL", rtp_bytes[:12]
+        )
+
+        if flags != 0b10000000:
+            raise RtpError("Padding and extension headers not supported")
+
+        payload_type &= 0x7F  # Remove marker bit
+        if payload_type != self.opus_payload:
+            raise RtpError(
+                f"Expected payload type {self.opus_payload}, got {payload_type}"
+            )
+
+        # Assume no padding, extension headers, etc.
+        opus_bytes = rtp_bytes[12:]
+
+        # Decode into raw audio.
+        # This will always be 48Khz stereo with 16-bit samples.
+        audio_bytes = opuslib.api.decoder.decode(
+            self._decoder,
+            opus_bytes,
+            len(opus_bytes),
+            self.opus_frame_size,
+            False,  # no forward error correction (fec)
+        )
+
+        # Convert to target sample rate, etc.
+        if channels == 1:
```

Comment:
Should this be `== 2`?

Reply:
No, this is the target number of channels.